### PR TITLE
otelhttp: Record metrics on timed out requests

### PR DESCRIPTION
**Pull request by carrbs**
fixes: #4542 

Copied verbatim the `withoutCancel` flow (referenced in #4542) and used it to ensure the metrics are recorded even if the parent `ctx` is closed. This "patch" (as well as the one in: `instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go`) can be updated to use the `context.WithoutContext` in Go 1.21 once we are no longer supporting Go 1.20.

### CODE REVIEW

**Review by pellared — COMMENTED** [review]
Implementation looks OK.

I tsuggest adding a test in `instrumentation/net/http/otelhttp/test/handler_test.go`.

I would try adding a server that would always have a canceled request. I guess that something like this should work:

```go
	// here setup a handler instrumented with otelhttp

	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	l, err := net.Listen("tcp", "localhost:0")
	require.NoError(t, err)
	srv := &http.Server{
		BaseContext:  func(_ net.Listener) context.Context { return ctx },
		ReadTimeout:  time.Second,
		WriteTimeout: 10 * time.Second,
		Handler:      otelHandler,
	}
	go func() {
		// When Shutdown is called, Serve immediately returns ErrServerClosed.
		assert.Equal(t, http.ErrServerClosed, srv.Serve(l))
	}()
	t.Cleanup(func() {
		assert.NoError(t, srv.Shutdown(context.Background()) 
	})

	// here make a call using HTTP Client. The address of the server is l.Addr().String()
```

**Comment by carrbs** [comment]
Hi @pellared, I attempted to test `ctxWithoutCancel` with your suggestion, but I think something is amiss either with my request or how I've instrumented the handler. The test passes whether or not `ctxWithoutCancel` is used. Likely (I hope) it's something obvious you can redirect me with on how I set up the test.

File: instrumentation/net/http/otelhttp/test/handler_test.go [hunk 1]
```diff
@@ -539,3 +541,64 @@ func TestWithRouteTag(t *testing.T) {
...
+
+	t.Cleanup(func() {
+		assert.NoError(t, srv.Shutdown(context.Background()))
+	})
+
+	req, err := http.NewRequest("GET", "http://"+l.Addr().String(), nil)
+	require.NoError(t, err)
+
+	rr := httptest.NewRecorder()
+	otelHandler.ServeHTTP(rr, req)
```
Comment (pellared) [inline instrumentation/net/http/otelhttp/test/handler_test.go:583]: This is not making a real HTTP request.

You should probably do something like this instead\

```suggestion
	resp, err := client.Do(req)
	require.NoError(t, err)
```

Comment (pellared) [inline instrumentation/net/http/otelhttp/test/handler_test.go:583]: This is not making a real HTTP request.

You should probably do something like this instead:

```suggestion
	_, err := client.Get("http://"+l.Addr().String())
	require.NoError(t, err)
```

File: instrumentation/net/http/otelhttp/handler.go [hunk 2]
```diff
@@ -281,3 +283,34 @@ func WithRouteTag(route string, h http.Handler) http.Handler {
 		h.ServeHTTP(w, r)
 	})
 }
+
+func withoutCancel(ctx context.Context) context.Context {
+	if ctx.Done() == nil {
+		return ctx
+	}
+	return withoutCancelCtx{ctx}
```
Comment (pellared) [inline instrumentation/net/http/otelhttp/handler.go:291]: Why not?

```suggestion
	if parent == nil {
		panic("cannot create context from nil parent")
	}
	return withoutCancelCtx{parent}
```

It is copied from https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/context/context.go;l=563-571 therefore I would keep it.

**Comment by pellared** [comment]
@carrbs Has mu suggestion helped?

File: instrumentation/net/http/otelhttp/handler.go (hunk 2 above)
Comment (carrbs) [reply instrumentation/net/http/otelhttp/handler.go:291]: Sounds good, I didn't know this had been copied from this. The thinking was that, if a context that wasn't cancellable was passed to `withoutCancel` it seemed like we could return/no-op (and, if it was `nil`, we'd get a panic).

**Comment by carrbs** [comment]
@pellared, yes, thank you! I'd overlooked that when trying to track down how to test this. It's passing now and I confirmed 
here: 
https://github.com/open-telemetry/opentelemetry-go/blob/6cee2b4a4c76b581115d0d0ca150ad8b2e683db6/sdk/metric/instrument.go#L208-L210
and here:
https://github.com/open-telemetry/opentelemetry-go/blob/6cee2b4a4c76b581115d0d0ca150ad8b2e683db6/sdk/metric/instrument.go#L208-L210
The expected behavior of the `withoutCancelCtx`. I'm gonna clean the PR up and take it out of draft.

File: instrumentation/net/http/otelhttp/handler.go [hunk 3]
```diff
@@ -232,13 +233,14 @@ func (h *middleware) serveHTTP(w http.ResponseWriter, r *http.Request, next http
 		attributes = append(attributes, semconv.HTTPStatusCode(rww.statusCode))
 	}
 	o := metric.WithAttributes(attributes...)
-	h.requestBytesCounter.Add(ctx, bw.read, o)
-	h.responseBytesCounter.Add(ctx, rww.written, o)
+	ctxWithoutCancel := withoutCancel(ctx)
```
Comment (carrbs) [inline instrumentation/net/http/otelhttp/handler.go:236]: Now that you've shared that [snippet from the context source](https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/context/context.go;l=563-571) @pellared, is there a reason we can't use that directly? I.e.
```suggestion
	ctxWithoutCancel := context.WithoutCancel(ctx)
```

(and then remove the local implementation)?

Comment (pellared) [reply instrumentation/net/http/otelhttp/handler.go:236]: Because `context.WithoutCancel` was introduced in Go 1.21 and we also support Go 1.20.

**Comment by pellared** [comment]
> I'm gonna clean the PR up and take it out of draft.

Please do not forget to add an entry in "Fixed" section in `CHANGELOG.md` 😉

File: CHANGELOG.md [hunk 4]
```diff
@@ -15,6 +15,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ### Fixed
 
+- Mimic the `context.WithoutCancel` behavior introduced in Go 1.21 to fix metric recording on timed out requests in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`. (#4542)
```
Comment (pellared) [inline CHANGELOG.md:18]:
```suggestion
- Fix `http.Handler` metric instrumentation in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` to record metrics on timed out requests. (#4542)
```

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
@carrbs Thank you for your contribution. Do not hesitate to leave suggestions that could make future contributions easier.

**Comment by pellared** [comment]
@Aneurysm9 @dmathieu PTAL as code owners.

**Comment by carrbs** [comment]
what's up with the `codecov/patch` failed test? Is it expecting some unit testing?

**Comment by pellared** [comment]
> what's up with the `codecov/patch` failed test? Is it expecting some unit testing?

It is fine. It just notifies that some new code is not covered by tests which is fine. See: https://app.codecov.io/gh/open-telemetry/opentelemetry-go-contrib/pull/4634?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=open-telemetry

**Comment by pellared** [comment]
The fix would not be needed if https://github.com/open-telemetry/opentelemetry-go/pull/4671 would be accepted (and merged).

File: instrumentation/net/http/otelhttp/handler.go [hunk 5]
```diff
@@ -281,3 +283,34 @@ func WithRouteTag(route string, h http.Handler) http.Handler {
 		h.ServeHTTP(w, r)
 	})
 }
+
+func withoutCancel(parent context.Context) context.Context {
```
[the remark this answers is not in the document]
Comment (carrbs) [reply instrumentation/net/http/otelhttp/handler.go:287]: Happy to do this, but it sounding like the better option might be the SDK native one (so long as we would not need the flexibility of calling those `.Add` and `.Record` methods in such a way where we want _different_ behavior based on the context's cancelled state)?

[the remark this answers is not in the document]
Comment (carrbs) [reply instrumentation/net/http/otelhttp/handler.go:291]: Added, in case we opt to merge this PR. Thanks @MrAlias!

**Comment by pellared** [comment]
I am changing to draft because of https://github.com/open-telemetry/opentelemetry-go/pull/4671

@carrbs Feel free to review the referenced PR.

**Comment by pellared** [comment]
Closing per https://github.com/open-telemetry/opentelemetry-go/pull/4671

**Pull request closed without merging**